README.md
## Introduction

**CMake updates for easier installation and use**

Original repo (version 2.1) at sourceforge: <http://freesteam.sourceforge.net/>

Credit goes to the original authors: John Pye, etc

Migration from sourceforge to github and use of CMake and SWIG, to make it compatible with modern Linux system (version 3.0).

Credit goes to Qingfeng Xia <https://github.com/qingfengxia/freesteam>

[License](LICENSE.txt): GPL as original release

### Prerequisites

+ CMake, mandatory. Package 'cmake' on Fedora and Ubuntu
+ GSL (gnu scientific library), mandatory. Package 'gsl-devel' on Fedora, 'libgsl-dev' on Ubuntu
+ python (optional), required if 'FREESTEAM_WITH_PYTHON=ON'. Package 'python-devel' on Fedora, 'python-dev' on Ubuntu
+ SWIG (optional), required if 'FREESTEAM_WITH_PYTHON=ON'. Package 'swig' on Fedora and Ubuntu
+ matplotlib (optional), for curves plotting. Package 'python-matplotlib' on Fedora and Ubuntu

### Overview of the features 

On the latest github version 3.0 (released in Nov 2019)

+ cmake updates (cmake version 3.14.2 used) for python detection and use 
+ Added a cmake option for the generation of the python interface (FREESTEAM_WITH_PYTHON) 
+ added targets 'make install' and 'make test' 
+ tested and used on Ubuntu 20.04 and Fedora 30 with gcc 9.0.1, swig3 and python3

[Readme of version 3.0](README-3.0.md)

[Readme of version 2.1](README-2.1.txt)

### Installation with CMake

```bash
mkdir build
cd build
cmake ../freesteam_src -DCMAKE_INSTALL_PREFIX=../freesteam_install -DFREESTEAM_WITH_PYTHON=ON
make
make test 
make install 
```

Version definition: version jumps from 3.0 to 3.1 to indicate updates



CHANGELOG.txt
FREESTEAM CHANGELOG
===================

3.1
30 Dec 2020
-----------

- cmake updates (version 3.14.2 used) for python detection and use (Michael Ndjinga)
- Added a cmake option for the generation of the python interface (FREESTEAM_WITH_PYTHON) (Michael Ndjinga)
- added targets 'make install' and 'make test' (Michael Ndjinga)

3.0
08 Nov 2019
-----------

- no new feature added, but cmake build system added (Qingfeng Xia)


2.1
11 Apr 2013
-----------

- Add surface tension calculation
- Added (T,s,p) surface plots (Rod Stephenson)
- Addes speed of sound calculation for Region 3 (Garrett Greet)
- GTK+ GUI (Carlo Tegano)
- Experimental Qt GUI (Carlo)
- Experimental PyQt GUI (Juan-Jose Gomera)
- Ability to access transport properties from ASCEND
- Experimental LabVIEW bindings (Andrew Young)
- Fixed bug #522 with sonames on Gentoo (Michal Gorny)


2.0
24 Jan 2009
-----------

- Complete rewrite in plain C
- Two-way solver now makes use of GSL, instead of being standalone.
- New APIs for C, Python and ASCEND
- Extended the use of 'reverse' equations e.g. for (p,h)
- Updated to new IAPWS-IF97 release documents.
- Included code for generation of (p,h) and other diagrams.
- Added Debian packaging meta-info (with multiple packages)
- freesteam-config package updated to new C API.
- Added README.txt file.
- Changed SONAME of libfreesteam.so to '.1.0' due to new API.

0.8.1
Monday, 15 October 2007
-----------------------

General
- Added some test cases to check viscosity correlation in test.py

For Windows users:
- Fixed bug in isinfnan.h (missing #include <math.h>), with thanks to
  Bruna Abreu Calfa.

0.8
Friday, 21 September 2007
-------------------------

General:
- Fixed errors in the conductivity correlation.
- Some methods in steamcalculator.h have had their parameters changed.
- Fixed incorrect interpolation of densities in the saturation region.
- Added ASCEND model to calculate moist air properties.
- Added support for a 'freesteam-config' program.
- Added 'mW/mK' to measurement.cpp for conductivity values.

0.7.2
Sunday, 19 August 2007
----------------------

General:
- libfreesteam is now a shared library/DLL.

For Windows users:
- ASCEND hook DLL renamed to freesteam_ascend.dll for use with
  current ASCEND (0.9.5.113 and newer)
- Static library has been changed to a shared library (DLL) on Windows
  to reduce duplication of code between Python module and core DLL.
- Changed name of Python module to _freesteam.pyd in line with
  the new requirements of Python 2.5.

For Linux users:
- Various improvements to RPM packaging.
- Improved detection of ASCEND at compile-time.

0.7.1
Monday, 12 March 2007
---------------------
An update for users of the Windows installer, primarily:
- Added missing header files in the Windows installer (woops)
- Added example C++ code plus Makefile/SConstruct for Windows users.
Also added a significant number of new function bindings for the
ASCEND interface.

0.7
Thursday, 18 January, 2007
--------------------------
Added a Windows binary installer that static library (.a) for use with 
MinGW GCC compiler, Python bindings ("from freesteam import *"), and
external functions for use with the ASCEND modelling environment.

0.6.2
Wednesday, 17 January, 2007
---------------------------
Correcting the conversion factor for psi, so: Measurement(1,"bar").to("psi")
Some enhancements to the ASCEND modelling environment bindings.

0.6.1
Friday, 7 July 2006
-------------------
Removing some binary files that crept into the last release.

0.6
Friday, 7 July 2006
-------------------
Switched to the 'SCons' build tool. Autotools no longer supported.
Fixed a memory leak during two-phase calculations
Added Python bindings to allow steam calculations from Python.
Added experimental bindings to the ASCEND modelling environment.
Added 'Measurement' class to allow interfacing 'Units' class with Python.

0.5.5
Tuesday, 13 December 2005
-------------------------
Fixed to compile under GCC 4.01
Fixed a very minor error in the IAPWS-95 correlation
Tightened tolerance on phir,phi0 (etc) test in IAPWS-95 correlation tests.


0.5.4
Thursday, 2 June 2005
---------------------
Minor fixes to Solver2 debug output
Minor fixes to EMSO hook library
Added diagram type p-u
Added new EMSO CalcObject for pipe friction
EMSO test models: pipe with pressure drop


0.5.3
Tuesday, 22 March 2005
----------------------
Added command line arguments to diagram.cli.cpp.
EMSO hook library can now report available methods if invalid method attemped.
EMSO test models: added pump curve, rankine cycle.


0.5.2
Saturday, 12 March 2005
-----------------------
Added two-way T,h solver
Property plots with log axes where required
Minor bug fixes


0.5.1
Tuesday, 8 March 2005
---------------------
Fixed typos in EMSO integration file


0.5
Monday, 7 March 2005
--------------------
Expanded two-way solver to u,v, p,h, p,T, p,u, p,s, T,s.
Changed to a tree directory structure
Can now build on MinGW
Integration with 'EMSO' simulation software mostly complete
Preliminary integration with 'EES' solver software started
Fixed a memory leak occuring with re-thrown exceptions


0.4
Wednesday, 23 February 2005
---------------------------
Expanded two-way solver to u,v (all regions), p,h, p,u
Added a 'first guess' option which can be used as a framework for integrating backwards correlations safely
Added a little CLI app (diagram.cli) to plot boundary curves on arbitrary axes via MATLAB
Made tolerance on SatCurve solver finer to correct Solver2 problems


0.3.1
Friday, 11 February 2005
------------------------
Ported to GCC 3.4: just minor changes


0.3 
Thursday, 10 February 2005
--------------------------
Added two-way solver routine (u,v) and a general framework for making others
Fixed bugs in the one-way solver
const-ified many methods
Added lots more tests
Changed the saturation curve correlation for T > T_REG1_REG3


0.2
--------------------------

python/tables.py
#!/usr/bin/python
# vim: set fileencoding=utf-8 :

# tables - a script that produces the data for Steam Tables using
# freesteam the GPL steam table library

## Copyright 2009 Grant Ingram, this program is distributed under the
## terms of the GNU General Public License
## Author e-mail: g.l.ingram@durham.ac.uk

##     This program is free software; you can redistribute it and/or
##     modify it under the terms of the GNU General Public License as
##     published by the Free Software Foundation; either version 2 of
##     the License, or (at your option) any later version.

##     This program is distributed in the hope that it will be useful,
##     but WITHOUT ANY WARRANTY; without even the implied warranty of
##     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##     GNU General Public License for more details.

##     You should have received a copy of the GNU General Public
##     License along with this program; if not, write to the Free
##     Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
##     Boston, MA 02110-1301, USA.

# Note if you have installed freesteam but get errors saying it can't
# find the library the following might help:
# LD_LIBRARY_PATH=/usr/local/lib
# export LD_LIBRARY_PATH

from freesteam import *
from pylab import *

Ttriple = 0.01 # Ideally this would be freesteam.TTRIPLE ?
ptriple = 0.00611

def temperature_table(T_range,caption,fout):

    fout.write(r"""\begin{landscape}
    \begin{table}
    \renewcommand{\tabcolsep}{0.6cm}
    \small
    \centering
    \begin{tabular}{ c c c c c c c c c c c }
      & & \multicolumn{2}{c}{Density} & \multicolumn{2}{c}{Internal energy} & \multicolumn{2}{c}{Specific enthalpy} & \multicolumn{2}{c}{Specific entropy}\\
      & & \multicolumn{2}{c}{$\unit{kg/m^3}$} & \multicolumn{2}{c}{$\unit{kJ/kg}$} & \multicolumn{2}{c}{$\unit{kJ/kg}$} & \multicolumn{2}{c}{$\unit{kJ/kg K}$} \\ \cline{3-4} 
      % & & \multicolumn{2}{c}{$\underline{\hphantom{aaaaaaaaa}}$}\\
    temp & pressure & Water & Steam \\
    $T \quad \unit{^{\circ}C} $ & $p \quad \unit{bar} $ & $\rho_f$ & $\rho_g$ & $u_f$ & $u_g$ & $h_f$ & $h_g$ & $s_f$ & $s_g$ \\[2pt]
    \hline
    \\[2pt]
    """)

    for T in T_range:
        W = region4_Tx(T+273.15,0) # water
        V = region4_Tx(T+273.15,1) # vapour

        fout.write(r"%3.2f & %4.3f & %4.1f & %4.3f & %4.1f & %4.1f & %4.1f & %4.1f & %4.3f & %4.3f \\" \
                   % (T,V.p/1e5,W.rho,V.rho,W.u/1000,V.u/1000,W.h/1000,V.h/1000,W.s/1000,V.s/1000))
        fout.write("\n")
        
    fout.write(r"\\[3pt]\hline")
    fout.write(r"\end{tabular}")
    fout.write("\n")
    fout.write(r"\caption{%s}" % caption)
    fout.write("\n")
    fout.write("""\end{table}
    \end{landscape}""")

def pressure_table(p_range,caption,fout):
    fout.write(r"""\begin{landscape}
    \begin{table}
    \centering
    \begin{tabular}{ c c c c c c c c c c c }
    $p$ & $T$ & $\rho_f$ & $\rho_g$ & $u_f$ & $u_g$ & $h_f$ & $h_g$ & $s_f$ & $s_g$ \\
    $[\unit{bar}]$ & $[\unit{^\circ C}]$ & $[\unit{kg/m^3}]$ & $[\unit{kg/m^3}]$ & $[\unit{kJ/kg}]$ & $[\unit{kJ/kg}]$ & $[\unit{kJ/kg}]$ & $[\unit{kJ/kg}]$ & $[\unit{kJ/kg K}]$ & $[\unit{kJ/kg K}]$ \\
    \hline
    """)
    for p in p_range:
        T = Tsat_p(p*1e5)
        T = T - 273.15
        W = region4_Tx(T+273.15,0) # water
        V = region4_Tx(T+273.15,1) # vapour

        fout.write(r"%4.3f & %3.2f & %4.1f & %4.3f & %4.1f & %4.1f & %4.1f & %4.1f & %4.3f & %4.3f \\" \
                   % (V.p/1e5,T,W.rho,V.rho,W.u/1000,V.u/1000,W.h/1000,V.h/1000,W.s/1000,V.s/1000))
        fout.write("\n")
        

    fout.write(r"\end{tabular}")
    fout.write("\n")
    fout.write(r"\caption{%s}" % caption)
    fout.write("\n")
    fout.write("""\end{table}
    \end{landscape}""")

def superheat_table(caption,variable,fout,p_range,t_range):
    fout.write(r"""\begin{landscape}
    \begin{table}
    \centering
    \begin{tabular}{ c c c c c c c c c c c c c c c c c c c c }
    """)

    # Do pressure header 
    fout.write(r"$p$ / [$bar$] ")
    for i in p_range:
        fout.write(r"& %4.1f " % i )
    fout.write(r"\\")

    fout.write("\n")
    # Do saturate temp header
    fout.write(r"$T_{sat}$ / [$^\circ C$] ")
    for i in p_range:
        fout.write(r"& %4.1f " % (Tsat_p(i*1e5) - 273.15) )
    fout.write(r"\\")
    fout.write("\n")
    fout.write(r"\hline")

    # Temperature header line
    if variable == "enthalpy":
        variable_label = r"Specific Enthalpy / [$\unit{kJ/kg}$]"
    elif variable == "entropy":
        variable_label = r"Specific Entropy / [$\unit{kJ/kg K}$]"
    elif variable == "density":
        variable_label = r"Density / [$\unit{kg/m^3}$]"
    elif variable == "internal":
        variable_label = r"Specific Internal Energy / [$\unit{kJ/kg}$]"
    else:
        print( "Option not recognised")
        raise SystemExit
    fout.write(r"$T$/[$^\circ C$] & \multicolumn{%d}{c}{%s} \\" % ((len(p_range)-1),variable_label))
    fout.write("\n")
    
    for i in t_range:
        fout.write(r" %3.0f " % i)
        for j in p_range:
            S = steam_pT(j*1e5,i+273.15)
            if variable == "enthalpy":
                data = round((S.h / 1000),2)
                fout.write(r"& %4.4g " %  data)
            elif variable == "entropy":
                data = round((S.s / 1000),3)
                fout.write(r"& %4.4g " %  data)
            elif variable == "density":
                data = round(S.rho,3)
                fout.write(r"& %4.4g " %  data)
            elif variable == "internal":
                data = round((S.u / 1000),2)
                fout.write(r"& %4.4g " %  data)
            else:
                print( "Option not recognised")
                raise SystemExit

        fout.write(r"\\")
        fout.write("\n")


    fout.write(r"\end{tabular}")
    fout.write("\n")
    fout.write(r"\caption{%s}" % caption)
    fout.write("\n")
    fout.write("""\end{table}
    \end{landscape}""")
    
#    $p$ & $T$ & $\rho_f$ & $\rho_g$ & $u_f$ & $u_g$ & $h_f$ & $h_g$ & $s_f$ & $s_g$ \\
#    $[bar]$ & $[^\circ C]$ & $[kg/m^3]$ & $[kg/m^3]$ & $[kJ/kg]$ & $[kJ/kg]$ & $[kJ/kg]$ & $[kJ/kg]$ & $[kJ/kg K]$ & $[kJ/kg K]$ \\
#    \hline
#    """)

def thermophys_table(caption, T_range):

	fout.write(r"""
	\begin{table}
	\centering
	\begin{tabular}{ c              c             c            c              c           c }
		             $T$          & $c_p$       & $\rho$     & $\mu$        & $k$       & $\mathrm{Pr}$ \\
		             $\mathrm{[^\circ C]}$ & $\mathrm{[kJ/kg \cdot K]}$ & $\mathrm{[kg/m^3]}$ & $\mathrm{[Pa \cdot s]}$ & $\mathrm{[W/m \cdot K]}$ & $$ \\
	\hline
	""")

	for T in T_range:
		S = steam_Tx(T + 273.15, 0)
	
		k = S.k
		mu = S.mu
		cp = S.cp
		Pr = cp * mu / k
		fout.write(r"%3.2f & %4.3f & %4.1f & %4.3e & %4.3f & %4.3f \\" \
			% (T, cp/1e3, S.rho, mu, k, Pr)
		)
		fout.write("\n")
		

	fout.write(r"\end{tabular}")
	fout.write("\n")
	fout.write(r"\caption{%s}" % caption)
	fout.write("\n")
	fout.write("""\end{table}\n""")

print( "tables.py - script to get data for Steam Tables")

# Create a latex file
fout = open("steamtable.tex","w")
fout.write(r"""\documentclass[a4paper,11pt]{article}
\usepackage{graphicx}
\usepackage{lscape}
%\usepackage[scaled=.92]{helvet}
\usepackage[hang, small, bf]{caption}
\usepackage{amssymb}
\usepackage{units}
\renewcommand{\familydefault}{\sfdefault}
\author{Scripted by Grant Ingram and John Pye}
\title{Steam Tables calculated using freesteam http://freesteam.sf.net}
\date{\today}
\begin{document}
\maketitle
""")

version_string = "Calculated with freesteam v%s using the IAPWS-IF97 Industrial Formulation \n \listoftables \n \listoffigures" % FREESTEAM_VERSION

fout.write(version_string)

print( "Creating Table 7...")
# Equivalent to Table 7 in Haywood Steam Tables: Saturated Water and
# Steam by Temperature from Triple Point to 100 deg C
# Actually we split into two tables 7a and 7b so as to make typesetting easier

#Table 7A
T_range = [Ttriple] + list(range(2,51,2)) # Range of table in deg C
caption = r"Saturated Water and Steam (Triple Point to 50$\unit{^\circ C}$)"
temperature_table(T_range,caption,fout)

#Table 7B
T_range = range(50,101,2) # Range of table in deg C
caption = r"Saturated Water and Steam (50 to 100 $\unit{^\circ C}$)"
temperature_table(T_range,caption,fout)

print( "Creating Table 8...")
#Table 8 - values by pressure all the way up to the crtical point....

# Given the non-uniform spacing here sometimes the manual way is the best...
p_range = (ptriple,0.008,0.01,0.012,0.014,0.016,0.018,0.020,0.022,0.024,0.026,0.028,0.030,0.035,0.040,0.045, \
           0.050,0.06,0.07,0.08,0.09,0.1,0.12,0.13,0.14,0.15)
caption = r"Saturated Water and Steam (0.00611 to 0.15 $\unit{bar}$)"
pressure_table(p_range,caption,fout)

#Table 8B
p_range = (0.16,0.18,0.20,0.22,0.24,0.26,0.28,0.30,0.35,0.40,0.45,0.50,0.55,0.60,0.65,0.70,0.75,0.80,0.85,0.90,0.95,1.00)
caption = r"Saturated Water and Steam (0.16 to 1 $\unit{bar}$)"
pressure_table(p_range,caption,fout)

#Table 8C
p_range = (1,1.1,1.2,1.3,1.4,1.5,1.6,1.7,1.8,1.9,2,2.2,2.4,2.6,2.8,3,3.2,3.4,3.6,3.8,4,4.2,4.4,4.6,4.8,5,5.5,6,6.5,7)
caption = r"Saturated Water and Steam (1 to 7 $\unit{bar}$)"
pressure_table(p_range,caption,fout)

#Table 8D
p_range = (7.5,8,8.5,9,9.5,10,10.5,11,11.5,12,12.5,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30)
caption = r"Saturated Water and Steam (7.5 to 30 $\unit{bar}$)"
pressure_table(p_range,caption,fout)

#Table 8D
p_range = (30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88)
caption = r"Saturated Water and Steam (30 to 88 $\unit{bar}$)"
pressure_table(p_range,caption,fout)

#Table 8D
p_range = (90,92,94,96,98,100,105,110,115,120,125,130,135,140,145,150,155,160,165,170,175,180,185,190,195,200,205,210,215,220)
caption = r"Saturated Water and Steam (90 to 220 $\unit{bar}$)"
pressure_table(p_range,caption,fout)

print( "Creating Table 9,10,11 and 12")
p_range = [0.1,0.5,1,5,10,20,40,60,80,100,150,200,250,300,400,500]
t_range = [0,25,50,75,100,125,150,175,200,225,250,275,300,350,400,450,500,550,600,650,700,750,800]
superheat_table("Enthalpy of Water and Steam","enthalpy",fout,p_range,t_range)
superheat_table("Entropy of Water and Steam","entropy",fout,p_range,t_range)
superheat_table("Density of Water and Steam","density",fout,p_range,t_range)
superheat_table("Internal Energy of Water and Steam","internal",fout,p_range,t_range)

print( "Creating thermophysical properties table")
t_range = [0,5,10,15,20,25,30,35,40,45,50,55,60,65,70,75,80,85,90,95,100,110,120,130,140,150,160,180,200,220,240,260,280,300,320,340,360, 370]
thermophys_table("Transport properties of water (saturated liquid)",t_range)

print( "including hsdiagram")
fout.write(r"""

\begin{figure}
\centering
\includegraphics[width=1.00\textwidth]{mollier}
\caption{Enthalpy-Entropy Diagram}
\label{fig:hsdiag}
\end{figure}

""")

fout.write("\end{document}")
fout.close
